src/libraries/System.Private.CoreLib/src/System/Text/Unicode/Utf8.cs

```diff
@@ -221,5 +222,414 @@ public static unsafe OperationStatus ToUtf16(ReadOnlySpan<byte> source, Span<cha
+                // exists purely to help make cases from (b) compile. Just delegate to the T-based implementation.
+                AppendFormatted<object?>(value, alignment, format);
+
+            /// <summary>Formats the value using the custom formatter from the provider.</summary>
+            /// <param name="value">The value to write.</param>
+            /// <param name="format">The format string.</param>
+            /// <typeparam name="T">The type of the value to write.</typeparam>
+            [MethodImpl(MethodImplOptions.NoInlining)]
+            private bool AppendCustomFormatter<T>(T value, string? format)
+            {
+                // This case is very rare, but we need to handle it prior to the other checks in case
+                // a provider was used that supplied an ICustomFormatter which wanted to intercept the particular value.
+                // We do the cast here rather than in the ctor, even though this could be executed multiple times per
+                // formatting, to make the cast pay for play.
+                Debug.Assert(_hasCustomFormatter);
+                Debug.Assert(_provider is not null);
+
+                ICustomFormatter? formatter = (ICustomFormatter?)_provider.GetFormat(typeof(ICustomFormatter));
+                Debug.Assert(formatter is not null, "An incorrectly written provider said it implemented ICustomFormatter, and then didn't");
+
+                if (formatter is not null &&
+                    formatter.Format(format, value, _provider) is string customFormatted)
+                {
+                    return AppendFormatted(customFormatted.AsSpan());
+                }
+
+                return true;
+            }
+
+            /// <summary>Writes the specified ISpanFormattable to the handler.</summary>
+            /// <param name="value">The value to write. It must be an ISpanFormattable but isn't constrained because the caller doesn't have a cosntraint.</param>
+            /// <param name="format">The format string.</param>
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            private bool AppendSpanFormattable<T>(T value, string? format)
+            {
+                Debug.Assert(value is ISpanFormattable);
+
+                Span<char> utf16 = stackalloc char[256];
+                return ((ISpanFormattable)value).TryFormat(utf16, out int charsWritten, format, _provider) ?
+                    AppendFormatted(utf16.Slice(0, charsWritten)) :
+                    GrowAndAppendFormatted(ref this, value, utf16.Length, out charsWritten, format);
+
+                [MethodImpl(MethodImplOptions.NoInlining)]
+                static bool GrowAndAppendFormatted(scoped ref TryWriteInterpolatedStringHandler thisRef, T value, int length, out int charsWritten, string? format)
+                {
+                    Debug.Assert(value is ISpanFormattable);
+
+                    while (true)
+                    {
+                        length *= 2;
```

Comment:
For reliability, cap this to `MaxLength`, like our other collection "double and retry" implementations do.

https://github.com/dotnet/runtime/blob/5711ca33f8dd1758e18c09ebdedf7e95eb424e0f/src/libraries/System.Private.CoreLib/src/System/IO/MemoryStream.cs#L146-L151

https://github.com/dotnet/runtime/blob/5711ca33f8dd1758e18c09ebdedf7e95eb424e0f/src/libraries/System.Private.CoreLib/src/System/Collections/Generic/List.cs#L451-L466